# Project 3 - Sentiment Analysis of Elon Musk using YouTube Comments

## Part 1 - Data Extraction, Cleaning and Exploratory Data Analysis (EDA)

## 1. Introduction and Problem Statement

Public image refers to the 'the character or attitudes that most people think' a person or organisation has ([source](https://www.ldoceonline.com/dictionary/public-image)). In most cases, a good public image is essential in:
1. Building trust, 
2. Attracting customers, and 
3. Fostering positive relationships with stakeholders.

As such, it is important for individuals and organisations to keep their finger on the pulse when it comes to understanding public sentiment around them at all times. In this project, I am going to focus on the public sentiment of a well-known and polarising figure the past few years, Elon Musk. In this scenario, I am part of a data science-backed PR agency to help Elon navigate the murky waters of media scrutiny and public perception, ensuring his actions and communications align with his goals.


The problem statement is:
**<center>Can we create an effective classification model to identify the level of negative sentiment in text comments?</center>**

The ideal outcome / goal is to help individuals and companies with understanding public sentiment around them in order to build their PR strategy. This will be done by creating a sentiment analysis model, with the dependent variable being a classifier whether a comment made about them is negative or not (where '1' refers to a negative comment and '0' refers to a non-negative comment).

Success of the model will be determined by the recall and accuracy scores of the model. The higher the recall and accuracy scores, the better it is at predicting the public sentiment around an individual or company. Between the two, we will prioritise the recall score in order to minimise missing negative sentiments as much as possible. This is so that we have more data to look through if we want to dive deeper into understanding the reasons behind negative sentiments.

The primary audience will be individuals and companies of interest, while the secondary audience will be  existing PR companies that we can explore partnerships with.

## 2. Data Extraction

In [1]:
import googleapiclient.discovery
import pandas as pd
import html
from bs4 import BeautifulSoup

from langdetect import detect

import spacy
import nltk
from nltk.corpus import stopwords
from nltk.tokenize import word_tokenize

import matplotlib.pyplot as plt
import seaborn as sns
from wordcloud import WordCloud, STOPWORDS, ImageColorGenerator
import re
import pickle

To train our model, we will be using comments from the YouTube video ['Joe Rogan Experience #1169 - Elon Musk'](https://www.youtube.com/watch?v=ycPr5-27vSI). This is because:
1. It is the most viewed video about Elon Musk (with approximately 67 million views, and 140,000 comments)
2. The video was created 4 years ago, so our model would be able to learn from a wide range of comments from 4 years ago until today.

Due to YouTube API limitations, we will pull 50,000 comments from this video, which should be more than enough to train our model.

In [2]:
# YouTube API Credentials
api_service_name = "youtube"
api_version = "v3"
DEVELOPER_KEY = "AIzaSyBGcph7UNs6FYit6VSrk986oKrqb8fEzpA"

In [3]:
# Using YouTube API to pull 50000 comments from a video
def get_video_comments(video_id, max_results=100, max_comments=50000):
    youtube = googleapiclient.discovery.build(api_service_name, api_version, developerKey=DEVELOPER_KEY)
    
    comments = []
    next_page_token = None
    total_comments_fetched = 0
    
    while total_comments_fetched < max_comments:
        remaining_comments_to_fetch = max_comments - total_comments_fetched
        comments_to_fetch = min(max_results, remaining_comments_to_fetch)
        
        request = youtube.commentThreads().list(
            part="snippet",
            videoId=video_id,
            maxResults=comments_to_fetch,
            pageToken=next_page_token if next_page_token else None
        )
        
        response = request.execute()
        
        # Get the video information (title and upload date) for the first batch of comments
        if total_comments_fetched == 0:
            video_response = youtube.videos().list(
                part="snippet",
                id=video_id
            ).execute()
            
            video_title = video_response['items'][0]['snippet']['title']
            video_published_at = video_response['items'][0]['snippet']['publishedAt']
        
        for item in response['items']:
            comment = item['snippet']['topLevelComment']['snippet']
            comments.append([
                comment['authorDisplayName'],
                comment['publishedAt'],
                comment['likeCount'],
                comment['textDisplay'],
                video_title,  # Adding video title for each comment
                video_published_at  # Adding video upload date for each comment
            ])
        
        total_comments_fetched += comments_to_fetch
        next_page_token = response.get('nextPageToken')
        if not next_page_token:
            break
    
    return comments

# List of video IDs for which you want to fetch comments
video_ids = ["ycPr5-27vSI"]

# Initialize an empty list to store all comments from multiple videos
all_comments = []

for video_id in video_ids:
    comments_for_video = get_video_comments(video_id)
    all_comments.extend(comments_for_video)

# Create a DataFrame from all the comments
df = pd.DataFrame(all_comments, columns=['author', 'published_at', 'like_count', 'text', 'video_title', 'video_published_at'])

# Display the first 10 rows of the DataFrame
df.head(10)

author          published_at  like_count  \
0              Ronnie Howard  2023-08-08T12:46:01Z           0   
1      Daham Kumarapathirana  2023-08-08T02:09:05Z           0   
2               Beau Johnson  2023-08-07T09:54:50Z           0   
3                ProphetKilo  2023-08-06T23:22:43Z           0   
4                   Khadulau  2023-08-06T22:35:45Z           0   
5               Preeti Singh  2023-08-06T10:55:29Z           0   
6            Overcomecross98  2023-08-05T21:55:21Z           2   
7                      Jesse  2023-08-05T16:27:07Z           0   
8                  supertiaj  2023-08-05T00:48:09Z           0   
9  Kilraven Wild Song Artist  2023-08-04T06:48:30Z           0   

                                                text  \
0    I see I ribbit playing out in the next 50 years   
1    67 Million Joe Rogan&#39;s Watched this video 😅   
2  Joe, giant carbon filters are called trees the...   
3  The Sun is not actually on fire in the same wa...   
4  If anyone cares to look, Elon Musk wiki still ...   
5  Correction with ACID THEORY: It was not Buddhi...   
6  I feel like he would be really annoying to tal...   
7  &quot;running an engine with no resistance&quo...   
8                I would not consent to this AI crap   
9  Amen guys. All we need is love, love, love. Gr...   

                              video_title    video_published_at  
0  Joe Rogan Experience #1169 - Elon Musk  2018-09-07T08:12:43Z  
1  Joe Rogan Experience #1169 - Elon Musk  2018-09-07T08:12:43Z  
2  Joe Rogan Experience #1169 - Elon Musk  2018-09-07T08:12:43Z  
3  Joe Rogan Experience #1169 - Elon Musk  2018-09-07T08:12:43Z  
4  Joe Rogan Experience #1169 - Elon Musk  2018-09-07T08:12:43Z  
5  Joe Rogan Experience #1169 - Elon Musk  2018-09-07T08:12:43Z  
6  Joe Rogan Experience #1169 - Elon Musk  2018-09-07T08:12:43Z  
7  Joe Rogan Experience #1169 - Elon Musk  2018-09-07T08:12:43Z  
8  Joe Rogan Experience #1169 - Elon Musk  2018-09-07T08:12:43Z  
9  Joe Rogan Experience #1169 - Elon Musk  2018-09-07T08:12:43Z

In [5]:
# Saving the file as an Original DataFrame so we do not need to run the above function many times
with open('../data/00_Original_DataFrame.pickle', 'wb') as file:
    pickle.dump(df, file)

In [6]:
# Load the file
with open('../data/00_Original_DataFrame.pickle', 'rb') as file:
    df = pickle.load(file)

In [7]:
# How many comments did we get?
df.shape

(49999, 6)

## 3. Data Cleaning

We will start off with data cleaning - focusing on null values, duplicated values, data types, HTML-encoded entities, and foreign language.

### 3.1 Handling Null Values

In [8]:
# Looking for columns with null values
df.isnull().sum()

author                0
published_at          0
like_count            0
text                  0
video_title           0
video_published_at    0
dtype: int64

There are no null values.

### 3.2 Handling Duplicated Values

In [9]:
# Looking for duplicated columns
df['text'].duplicated().sum()

947

In [10]:
df[df['text'].duplicated()]

author          published_at  like_count  \
46                SolaraProject  2023-07-29T09:41:53Z           0   
711                       Maria  2023-01-20T05:38:18Z           0   
788             Loan Wroblewski  2022-12-30T00:04:10Z           3   
1575           Güneş türkyılmaz  2022-11-15T07:21:12Z           0   
1696                They Wright  2022-11-09T20:28:21Z           0   
...                         ...                   ...         ...   
49581                   Izzy G.  2018-10-12T19:08:31Z           0   
49704  Macbeth and her demon 🥀👹  2018-10-10T19:24:31Z           0   
49932               Mayne Maybe  2018-10-07T11:18:07Z           0   
49944                   Levi H.  2018-10-07T06:57:14Z           0   
49996             franc zamudio  2018-10-06T18:50:03Z           0   

                                                    text  \
46     Elon: &quot;Be nicer to each other&quot;<br>Jo...   
711    So what’s the statue of limitations on capital...   
788    <a href="https://www.youtube.com/watch?v=ycPr5...   
1575                                                  💯💯   
1696             You two have smoked way too much weed..   
...                                                  ...   
49581  <a href="https://www.youtube.com/watch?v=ycPr5...   
49704                                            Big pit   
49932                                                Wtf   
49944                                    Business magnet   
49996                                             Genius   

                                  video_title    video_published_at  
46     Joe Rogan Experience #1169 - Elon Musk  2018-09-07T08:12:43Z  
711    Joe Rogan Experience #1169 - Elon Musk  2018-09-07T08:12:43Z  
788    Joe Rogan Experience #1169 - Elon Musk  2018-09-07T08:12:43Z  
1575   Joe Rogan Experience #1169 - Elon Musk  2018-09-07T08:12:43Z  
1696   Joe Rogan Experience #1169 - Elon Musk  2018-09-07T08:12:43Z  
...                                       ...                   ...  
49581  Joe Rogan Experience #1169 - Elon Musk  2018-09-07T08:12:43Z  
49704  Joe Rogan Experience #1169 - Elon Musk  2018-09-07T08:12:43Z  
49932  Joe Rogan Experience #1169 - Elon Musk  2018-09-07T08:12:43Z  
49944  Joe Rogan Experience #1169 - Elon Musk  2018-09-07T08:12:43Z  
49996  Joe Rogan Experience #1169 - Elon Musk  2018-09-07T08:12:43Z  

[947 rows x 6 columns]

There are 947 rows where the comment is duplicated. Let us drop them (while keeping the first occurence).

In [11]:
df.drop_duplicates(subset='text', keep='first', inplace=True)

### 3.3 Handling Data Types

In [12]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 49052 entries, 0 to 49998
Data columns (total 6 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   author              49052 non-null  object
 1   published_at        49052 non-null  object
 2   like_count          49052 non-null  int64 
 3   text                49052 non-null  object
 4   video_title         49052 non-null  object
 5   video_published_at  49052 non-null  object
dtypes: int64(1), object(5)
memory usage: 2.6+ MB


We have two columns that are meant to be datetimes:
1. 'published_at'
2. 'video_published_at'

Let us change the type.

In [13]:
df['published_at'] = pd.to_datetime(df['published_at'])
df['video_published_at'] = pd.to_datetime(df['video_published_at'])

Additionally, the 'like_count' should be an integer. Let us investigate.

In [14]:
df['like_count'].unique()

array([    0,     2,     1,     3,     4,    14,     9,    12,    10,
         142,     6,     7,    26,    20,     8,    31,  1019,    11,
           5,    99,    33,    97,    17,    13,   700,    27,   164,
          86,    19,   388,    68,    47,   277,    18,    37,   107,
          25,    42,    65,    58,    41,   160,    70,   281,    28,
         505,    55,    30,    21,    59,   439,    15,    46,   199,
          62,   376,   829,    94,   206,    52,    40,    64,    88,
          60,    44,   136,  1436,    35,   244,    23,    73,    32,
         131,    76,   621,   146,   370,   139,    16,   280,   116,
          34,   898,    24,    50,   144,   121,    22,  1717,   519,
          91,   375,   323,    84,   343,   532,  1439,   127,   129,
         190,   340,  1486,   620,    79,    45,   231,   655,    29,
          49,    36,    63,    38,    57,    71,   237,    48,   265,
         622,   883,   103,    39,   102,   378,    89,   156,   273,
          74,   204,

There does not seem to be any issue with the column, so we can change its type to integer.

In [15]:
df['like_count'].astype(int)

0          0
1          0
2          0
3          0
4          0
        ... 
49993      0
49994    156
49995      0
49997      0
49998      0
Name: like_count, Length: 49052, dtype: int64

In [16]:
# Final data types
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 49052 entries, 0 to 49998
Data columns (total 6 columns):
 #   Column              Non-Null Count  Dtype              
---  ------              --------------  -----              
 0   author              49052 non-null  object             
 1   published_at        49052 non-null  datetime64[ns, UTC]
 2   like_count          49052 non-null  int64              
 3   text                49052 non-null  object             
 4   video_title         49052 non-null  object             
 5   video_published_at  49052 non-null  datetime64[ns, UTC]
dtypes: datetime64[ns, UTC](2), int64(1), object(3)
memory usage: 2.6+ MB


### 3.4 Handling HTML-encoded entities

As we saw earlier, the comments can be quite confusing due to HTML-encoded entities. Thus, we will use BeautifulSoup to handle that.

In [17]:
df['text'] = df['text'].apply(lambda x: BeautifulSoup(x, "html.parser").get_text())
df

/var/folders/w4/0rkrlv1d2cg4g7rbby5yb_tr0000gn/T/ipykernel_86881/2748582876.py:1: MarkupResemblesLocatorWarning: The input looks more like a filename than markup. You may want to open this file and pass the filehandle into Beautiful Soup.
  df['text'] = df['text'].apply(lambda x: BeautifulSoup(x, "html.parser").get_text())


author              published_at  like_count  \
0              Ronnie Howard 2023-08-08 12:46:01+00:00           0   
1      Daham Kumarapathirana 2023-08-08 02:09:05+00:00           0   
2               Beau Johnson 2023-08-07 09:54:50+00:00           0   
3                ProphetKilo 2023-08-06 23:22:43+00:00           0   
4                   Khadulau 2023-08-06 22:35:45+00:00           0   
...                      ...                       ...         ...   
49993           Arm Wrestler 2018-10-06 19:01:01+00:00           0   
49994                EL34XYZ 2018-10-06 18:54:27+00:00         156   
49995                  River 2018-10-06 18:52:02+00:00           0   
49997              Gaz Potts 2018-10-06 18:42:25+00:00           0   
49998       The Prime Remedy 2018-10-06 18:37:32+00:00           0   

                                                    text  \
0        I see I ribbit playing out in the next 50 years   
1            67 Million Joe Rogan's Watched this video 😅   
2      Joe, giant carbon filters are called trees the...   
3      The Sun is not actually on fire in the same wa...   
4      If anyone cares to look, Elon Musk wiki still ...   
...                                                  ...   
49993                  I rather be pessimistic and WRONG   
49994  Wow, jaw dropping interview. It's fascinating ...   
49995  A day later joe “Elon weird, I don’t understan...   
49997  I watched all this earlier I couldn't stop it ...   
49998       Lol at the description.  Elon got his magnet   

                                  video_title        video_published_at  
0      Joe Rogan Experience #1169 - Elon Musk 2018-09-07 08:12:43+00:00  
1      Joe Rogan Experience #1169 - Elon Musk 2018-09-07 08:12:43+00:00  
2      Joe Rogan Experience #1169 - Elon Musk 2018-09-07 08:12:43+00:00  
3      Joe Rogan Experience #1169 - Elon Musk 2018-09-07 08:12:43+00:00  
4      Joe Rogan Experience #1169 - Elon Musk 2018-09-07 08:12:43+00:00  
...                                       ...                       ...  
49993  Joe Rogan Experience #1169 - Elon Musk 2018-09-07 08:12:43+00:00  
49994  Joe Rogan Experience #1169 - Elon Musk 2018-09-07 08:12:43+00:00  
49995  Joe Rogan Experience #1169 - Elon Musk 2018-09-07 08:12:43+00:00  
49997  Joe Rogan Experience #1169 - Elon Musk 2018-09-07 08:12:43+00:00  
49998  Joe Rogan Experience #1169 - Elon Musk 2018-09-07 08:12:43+00:00  

[49052 rows x 6 columns]

### 3.5 Handling Foreign Languages

Finally, we will look to handle foreign languages.

In [18]:
# Creating an 'is_english' column
def is_english(text):
    try:
        lang = detect(text)
        return 1 if lang == 'en' else 0
    except:
        return 1

df['is_english'] = df['text'].apply(is_english)

In [19]:
df

author              published_at  like_count  \
0              Ronnie Howard 2023-08-08 12:46:01+00:00           0   
1      Daham Kumarapathirana 2023-08-08 02:09:05+00:00           0   
2               Beau Johnson 2023-08-07 09:54:50+00:00           0   
3                ProphetKilo 2023-08-06 23:22:43+00:00           0   
4                   Khadulau 2023-08-06 22:35:45+00:00           0   
...                      ...                       ...         ...   
49993           Arm Wrestler 2018-10-06 19:01:01+00:00           0   
49994                EL34XYZ 2018-10-06 18:54:27+00:00         156   
49995                  River 2018-10-06 18:52:02+00:00           0   
49997              Gaz Potts 2018-10-06 18:42:25+00:00           0   
49998       The Prime Remedy 2018-10-06 18:37:32+00:00           0   

                                                    text  \
0        I see I ribbit playing out in the next 50 years   
1            67 Million Joe Rogan's Watched this video 😅   
2      Joe, giant carbon filters are called trees the...   
3      The Sun is not actually on fire in the same wa...   
4      If anyone cares to look, Elon Musk wiki still ...   
...                                                  ...   
49993                  I rather be pessimistic and WRONG   
49994  Wow, jaw dropping interview. It's fascinating ...   
49995  A day later joe “Elon weird, I don’t understan...   
49997  I watched all this earlier I couldn't stop it ...   
49998       Lol at the description.  Elon got his magnet   

                                  video_title        video_published_at  \
0      Joe Rogan Experience #1169 - Elon Musk 2018-09-07 08:12:43+00:00   
1      Joe Rogan Experience #1169 - Elon Musk 2018-09-07 08:12:43+00:00   
2      Joe Rogan Experience #1169 - Elon Musk 2018-09-07 08:12:43+00:00   
3      Joe Rogan Experience #1169 - Elon Musk 2018-09-07 08:12:43+00:00   
4      Joe Rogan Experience #1169 - Elon Musk 2018-09-07 08:12:43+00:00   
...                                       ...                       ...   
49993  Joe Rogan Experience #1169 - Elon Musk 2018-09-07 08:12:43+00:00   
49994  Joe Rogan Experience #1169 - Elon Musk 2018-09-07 08:12:43+00:00   
49995  Joe Rogan Experience #1169 - Elon Musk 2018-09-07 08:12:43+00:00   
49997  Joe Rogan Experience #1169 - Elon Musk 2018-09-07 08:12:43+00:00   
49998  Joe Rogan Experience #1169 - Elon Musk 2018-09-07 08:12:43+00:00   

       is_english  
0               1  
1               1  
2               1  
3               1  
4               1  
...           ...  
49993           1  
49994           1  
49995           1  
49997           1  
49998           1  

[49052 rows x 7 columns]

In [20]:
df['is_english'].value_counts()

1    44315
0     4737
Name: is_english, dtype: int64

In [21]:
df[df['is_english']==0].head(50)

author              published_at  like_count  \
18                           Cooper6006 2023-08-02 04:53:26+00:00           0   
21                           Cooper6006 2023-08-02 04:38:09+00:00           0   
27                      David McFarland 2023-08-02 02:27:04+00:00           0   
30                           YELLOW  FF 2023-08-01 18:04:21+00:00           0   
42                         Junhao Liang 2023-07-29 23:41:37+00:00           0   
49                                  Kid 2023-07-28 02:10:04+00:00           0   
57                         Gabe Moralez 2023-07-24 18:10:13+00:00           0   
58                            ZeeHamada 2023-07-24 03:02:19+00:00           0   
104                            On Point 2023-07-07 04:46:18+00:00           0   
125                     gbeatsmacedonia 2023-06-26 11:13:29+00:00           0   
126              Cassie Kay Cronemiller 2023-06-26 09:11:09+00:00           0   
128                              Artizt 2023-06-26 00:27:47+00:00           0   
149                        class-d noob 2023-06-18 13:05:36+00:00           1   
169                    Daniel Papendorf 2023-06-17 06:36:09+00:00           0   
172                    Daniel Papendorf 2023-06-17 06:31:21+00:00           0   
174                    Daniel Papendorf 2023-06-17 06:30:18+00:00           0   
175                    Daniel Papendorf 2023-06-17 06:29:52+00:00           0   
178                    Daniel Papendorf 2023-06-17 06:28:12+00:00           0   
181       AKASH#RDJ feeling transformer 2023-06-16 18:54:15+00:00           1   
191                         Chady Aniki 2023-06-13 08:49:57+00:00           0   
200                             S Davis 2023-06-11 03:18:56+00:00           1   
201                     Free Minded Ent 2023-06-11 02:18:10+00:00           0   
207          HealthyHorsez BOOK INSIDER 2023-06-08 17:29:22+00:00           0   
208                        David Fowler 2023-06-08 04:51:38+00:00           0   
224                       ŅoĆtRŅlcr3@Țř 2023-06-02 02:30:33+00:00           0   
237                               Benz0 2023-05-28 23:21:01+00:00           0   
246                         LUIS MADRID 2023-05-25 21:55:35+00:00           0   
263                      Michael Watson 2023-05-20 10:19:39+00:00           0   
298         The Richest YouTube channel 2023-05-10 09:58:15+00:00           0   
306                        Justin Quick 2023-05-08 04:31:15+00:00           0   
309                       MarcussiousX🤔 2023-05-07 03:34:35+00:00           0   
310                         Tokyo Shemp 2023-05-07 02:48:12+00:00           1   
327                       Thomas Butryn 2023-05-01 19:55:19+00:00           0   
331                             Anes P. 2023-04-29 17:31:34+00:00           0   
334                           Ian Brand 2023-04-28 04:56:03+00:00           0   
339                       Peter Scholtz 2023-04-27 02:05:55+00:00           0   
350  Firulay Antonio Muñoz Victor Muñoz 2023-04-22 21:39:54+00:00           0   
351                                 Jay 2023-04-22 16:52:06+00:00           0   
373                            pay_attn 2023-04-14 18:23:25+00:00           0   
422                         Raghav Noob 2023-03-28 04:21:02+00:00           0   
435                     Trần Ngọc Quang 2023-03-23 07:33:10+00:00           0   
458                     Jennifer Copple 2023-03-18 20:40:31+00:00           0   
462                         Hynek Žižka 2023-03-18 16:54:25+00:00           0   
495                             Chuittv 2023-03-07 18:06:49+00:00           0   
496                        TONIC WATERS 2023-03-06 21:47:53+00:00           0   
504                       Burke:Charlie 2023-03-03 00:19:55+00:00           0   
511                     Oliveira Miguel 2023-02-28 20:10:07+00:00           0   
523                           BRUXO CFM 2023-02-25 01:17:13+00:00           0   
537                      shashibushan k 2023-02-22 10:37:52+00:0

We actually see that, based on the first 50 rows, a good majority of the comments are in English, with only a couple of rows being Spanish and Portugese. There are, however, a number of rows simply calling out timestamps in videos, YouTube links, as well as slang.

If we were to drop these, we still have over 44,000 comments to train our model. So, let us drop them.

In [22]:
df.drop(df[df['is_english'] == 0].index, inplace = True)
df.reset_index(drop = True, inplace = True)
df

author              published_at  like_count  \
0              Ronnie Howard 2023-08-08 12:46:01+00:00           0   
1      Daham Kumarapathirana 2023-08-08 02:09:05+00:00           0   
2               Beau Johnson 2023-08-07 09:54:50+00:00           0   
3                ProphetKilo 2023-08-06 23:22:43+00:00           0   
4                   Khadulau 2023-08-06 22:35:45+00:00           0   
...                      ...                       ...         ...   
44310           Arm Wrestler 2018-10-06 19:01:01+00:00           0   
44311                EL34XYZ 2018-10-06 18:54:27+00:00         156   
44312                  River 2018-10-06 18:52:02+00:00           0   
44313              Gaz Potts 2018-10-06 18:42:25+00:00           0   
44314       The Prime Remedy 2018-10-06 18:37:32+00:00           0   

                                                    text  \
0        I see I ribbit playing out in the next 50 years   
1            67 Million Joe Rogan's Watched this video 😅   
2      Joe, giant carbon filters are called trees the...   
3      The Sun is not actually on fire in the same wa...   
4      If anyone cares to look, Elon Musk wiki still ...   
...                                                  ...   
44310                  I rather be pessimistic and WRONG   
44311  Wow, jaw dropping interview. It's fascinating ...   
44312  A day later joe “Elon weird, I don’t understan...   
44313  I watched all this earlier I couldn't stop it ...   
44314       Lol at the description.  Elon got his magnet   

                                  video_title        video_published_at  \
0      Joe Rogan Experience #1169 - Elon Musk 2018-09-07 08:12:43+00:00   
1      Joe Rogan Experience #1169 - Elon Musk 2018-09-07 08:12:43+00:00   
2      Joe Rogan Experience #1169 - Elon Musk 2018-09-07 08:12:43+00:00   
3      Joe Rogan Experience #1169 - Elon Musk 2018-09-07 08:12:43+00:00   
4      Joe Rogan Experience #1169 - Elon Musk 2018-09-07 08:12:43+00:00   
...                                       ...                       ...   
44310  Joe Rogan Experience #1169 - Elon Musk 2018-09-07 08:12:43+00:00   
44311  Joe Rogan Experience #1169 - Elon Musk 2018-09-07 08:12:43+00:00   
44312  Joe Rogan Experience #1169 - Elon Musk 2018-09-07 08:12:43+00:00   
44313  Joe Rogan Experience #1169 - Elon Musk 2018-09-07 08:12:43+00:00   
44314  Joe Rogan Experience #1169 - Elon Musk 2018-09-07 08:12:43+00:00   

       is_english  
0               1  
1               1  
2               1  
3               1  
4               1  
...           ...  
44310           1  
44311           1  
44312           1  
44313           1  
44314           1  

[44315 rows x 7 columns]

## 4. Feature Engineering

To get more details of the comments, let us create a couple of columns: comment length and comment word count.

In [23]:
df['comment_length'] = df['text'].apply(lambda x: len(x))

In [24]:
df['comment_word_count'] = df['text'].apply(lambda x: len(x.split()))

In [25]:
df

author              published_at  like_count  \
0              Ronnie Howard 2023-08-08 12:46:01+00:00           0   
1      Daham Kumarapathirana 2023-08-08 02:09:05+00:00           0   
2               Beau Johnson 2023-08-07 09:54:50+00:00           0   
3                ProphetKilo 2023-08-06 23:22:43+00:00           0   
4                   Khadulau 2023-08-06 22:35:45+00:00           0   
...                      ...                       ...         ...   
44310           Arm Wrestler 2018-10-06 19:01:01+00:00           0   
44311                EL34XYZ 2018-10-06 18:54:27+00:00         156   
44312                  River 2018-10-06 18:52:02+00:00           0   
44313              Gaz Potts 2018-10-06 18:42:25+00:00           0   
44314       The Prime Remedy 2018-10-06 18:37:32+00:00           0   

                                                    text  \
0        I see I ribbit playing out in the next 50 years   
1            67 Million Joe Rogan's Watched this video 😅   
2      Joe, giant carbon filters are called trees the...   
3      The Sun is not actually on fire in the same wa...   
4      If anyone cares to look, Elon Musk wiki still ...   
...                                                  ...   
44310                  I rather be pessimistic and WRONG   
44311  Wow, jaw dropping interview. It's fascinating ...   
44312  A day later joe “Elon weird, I don’t understan...   
44313  I watched all this earlier I couldn't stop it ...   
44314       Lol at the description.  Elon got his magnet   

                                  video_title        video_published_at  \
0      Joe Rogan Experience #1169 - Elon Musk 2018-09-07 08:12:43+00:00   
1      Joe Rogan Experience #1169 - Elon Musk 2018-09-07 08:12:43+00:00   
2      Joe Rogan Experience #1169 - Elon Musk 2018-09-07 08:12:43+00:00   
3      Joe Rogan Experience #1169 - Elon Musk 2018-09-07 08:12:43+00:00   
4      Joe Rogan Experience #1169 - Elon Musk 2018-09-07 08:12:43+00:00   
...                                       ...                       ...   
44310  Joe Rogan Experience #1169 - Elon Musk 2018-09-07 08:12:43+00:00   
44311  Joe Rogan Experience #1169 - Elon Musk 2018-09-07 08:12:43+00:00   
44312  Joe Rogan Experience #1169 - Elon Musk 2018-09-07 08:12:43+00:00   
44313  Joe Rogan Experience #1169 - Elon Musk 2018-09-07 08:12:43+00:00   
44314  Joe Rogan Experience #1169 - Elon Musk 2018-09-07 08:12:43+00:00   

       is_english  comment_length  comment_word_count  
0               1              47                  11  
1               1              43                   8  
2               1              69                  13  
3               1             450                  81  
4               1              95                  17  
...           ...             ...                 ...  
44310           1              33                   6  
44311           1             328                  55  
44312           1              67                  13  
44313           1             186                  37  
44314           1              44                   8  

[44315 rows x 9 columns]

In [26]:
# Saving the file as pickle
with open('../data/01_Cleaned_Data.pickle', 'wb') as file:
    pickle.dump(df, file)